Refs #10162 - adds ability to diff configurations

mbacovsky opened the pull request:
- refactoring
- moved configuration specific logic from KafoConfigure to Configuration object
- ability to have multiple Configuration instances
- configuration merging
- configuration diff printing
- documentation

Tests are in progress and will be added in separate commit.

mbacovsky commented:
- small refactoring and typo fixes
- tests for scenario diffs added

ares commented on lib/kafo/scenario_manager.rb:
```diff
@@ -29,17 +29,18 @@ def available_scenarios
 
     def list_available_scenarios
-      say "Available scenarios"
+      say ::HighLine.color("Available scenarios", :cyan)
```
please use [custom defined colors](https://github.com/theforeman/kafo/blob/master/lib/kafo/color_scheme.rb#L34) in the whole PR, the reason is that colors are changed according to color scheme (black vs white background)

feel free to add new key to color hash when needed...

ares commented on lib/kafo/scenario_manager.rb:
```diff
@@ -98,23 +116,50 @@ def scenario_from_args
     end
 
+    def show_scenario_diff(prev_scenario, new_scenario)
+      say ::HighLine.color("Scenarios are being compared, that may take a while...", :cyan)
```
colors (see above)

ares commented on lib/kafo/scenario_manager.rb:
```diff
@@ -29,17 +29,18 @@ def available_scenarios
-        say "  #{content[:name]} (use: --scenario #{scenario})"
-        say "        " + content[:description] unless (content[:description].nil? || content[:description].empty?)
+        use = (config_file == @previous_scenario ? 'INSTALLED' : "use: --scenario #{scenario}")
+        say ::HighLine.color("  #{content[:name]} ", :yellow)
```
colors (see above)

ares commented on lib/kafo/scenario_manager.rb:
```diff
@@ -98,23 +116,50 @@ def scenario_from_args
+      missing = new_conf.params_missing(prev_conf)
+      changed = new_conf.params_changed(prev_conf)
+
+      say ::HighLine.color("\nValues from previous installation that will be added by installer:", :yellow)
```
colors (see above)

ares replied:
please avoid adding \n as part of HighLine styling, it would cause problems if we set some style to use background color

ares commented on lib/kafo/scenario_manager.rb:
```diff
@@ -98,23 +116,50 @@ def scenario_from_args
-        end
+        changed.each { |param| say "  #{param.module.class_name}::#{param.name}: #{param.value} -> #{prev_conf.param(param.module.class_name, param.name).value}" }
+      end
+      say ::HighLine.color("\nValues from previous installation that will be lost by scenario change:", :yellow)
```
colors (see above)

ares commented on test/kafo/configuration_test.rb:
```diff
@@ -0,0 +1,73 @@
+        cfg = { :modules_dir => './my_modules', :answer_file => 'test/fixtures/basic_answers.yaml'}
+        config_file = ConfigFileFactory.build('modules_dir', cfg.to_yaml).path
+        config = Kafo::Configuration.new(config_file, false)
+        assert_equal [File.join(current_dir, 'my_modules' )], config.module_dirs
```
nit, extra space before `)`

ares commented on README.md:
````diff
@@ -236,6 +236,43 @@ to look for module parameters. There are multiple ways how the installer can sel
+
+Installer tries to prevent unintentional change of a scenario and interrupts when such situation is detected:
+```bash
+  $> foreman-installer -S foreman-installer
````
other examples are either with `$` or without anything, please choose one to not introduce even more inconsistency

ares commented on README.md:
````diff
@@ -236,6 +236,43 @@ to look for module parameters. There are multiple ways how the installer can sel
+To avoid losing some configuration values installer can detect differences between answer files of the two scenarios.
+To display them use either interactive mode (`-i`) or `--compare-scenarios` flag:
+```bash
+  $> foreman-installer --compare-scenarios --scenario foreman
````
same as above

ares commented on lib/kafo/scenario_manager.rb:
```diff
@@ -29,17 +29,18 @@ def available_scenarios
+        use = (config_file == @previous_scenario ? 'INSTALLED' : "use: --scenario #{scenario}")
+        say ::HighLine.color("  #{content[:name]} ", :yellow)
+        say "(#{use})"
+        say "        " + content[:description] if content[:description]
```
this break the listing when description is empty string
![kafo](https://cloud.githubusercontent.com/assets/109773/11956877/d5c1be0c-a8bd-11e5-907c-179d0733449f.png)

ares commented on lib/kafo/scenario_manager.rb:
```diff
@@ -29,17 +29,18 @@ def available_scenarios
         scenario = File.basename(config_file, '.yaml')
-        say "  #{content[:name]} (use: --scenario #{scenario})"
-        say "        " + content[:description] unless (content[:description].nil? || content[:description].empty?)
+        use = (config_file == @previous_scenario ? 'INSTALLED' : "use: --scenario #{scenario}")
```
detection does not work when relative path is used, I guess `File.expand_path` should be used

ares commented:
when comparing scenarios, I'm getting the trace 

```
/home/ares/.rbenv/versions/2.0.0-p247/gemsets/kafo1/gems/kafo-0.6.13/lib/kafo/params/password.rb:54:in `phrase': undefined method `app' for nil:NilClass (NoMethodError)
```

looking at https://github.com/theforeman/kafo/blob/master/lib/kafo/params/password.rb#L54 - we need to get configuration instance to the param

ares commented:
Will this change require a change in https://github.com/theforeman/foreman-installer/blob/42521b79e4e4a7a05dbf30b6fd5a2b783fb2e871/bin/foreman-installer#L21 ? I suppose that should be ok, since it's after the run so the config would be associated already...

ares commented on lib/kafo/scenario_manager.rb:
```diff
@@ -98,23 +116,50 @@ def scenario_from_args
+      end
+      say ::HighLine.color("\nValues from previous installation that will be lost by scenario change:", :yellow)
+      if missing.empty?
+        say " No values from previous installation will be lost"
```
one space indentation, while few lines above you used 2

ares commented on lib/kafo/scenario_manager.rb:
```diff
@@ -98,23 +116,50 @@ def scenario_from_args
-          message = "You are trying to replace existing installation with different scenario. This may lead to unpredictable states. Use --force to override."
-          fail_now(message,:scenario_error)
-        end
+        changed.each { |param| say "  #{param.module.class_name}::#{param.name}: #{param.value} -> #{prev_conf.param(param.module.class_name, param.name).value}" }
```
I'd expect opposite order $previous -> $new_value

mbacovsky replied:
I agree this is confusing. Currently it states that:
$new_default_value will be replaced with $previous as it may be custom value, Thus '$new -> $previous' made better sense to me. How about changing the description to: `Defaults that will be updated with values from previous installation`?

ares replied:
yeah that sounds better

ares commented on lib/kafo/configuration.rb:
```diff
@@ -151,17 +171,47 @@ def params
+    def params_missing(old_config)
+      # finds params that are present but will be missing in the new config
+      old_config.params.select do |par|
+        param(par.module.class_name, par.name).nil?
```
what would be the example here? if it's the same module, the parameter is always there and gets the default value

mbacovsky replied:
This will cover situation, when module present in previous is missing from new scenario. All affected parameters from previous will be listed. E.g. I had proxy installed and wanted to reinstall full foreman scenario. There were some proxy plugins missing from the new scenario and this warned me.

ares replied:
ok, that make sense, e.g. template plugin can be missing when proxy is installed with foreman, thanks for clarification

ares commented:
suppose in my old scenario answer file I have following 
```
foreman: false
```

while in new
```
foreman:
  foreman_url: something
```

I'm getting `foreman::foreman_url: ["something"] -> ["https://edna.brq.redhat.com"]` which is probably not expected. I guess we should check whether module is enabled (https://github.com/theforeman/kafo/blob/master/lib/kafo/puppet_module.rb#L36)

also the opposite situation gives me `foreman::foreman_url: ["https://edna.brq.redhat.com"] -> ["something"]`

mbacovsky commented:
I've updated the PR, testing it.
- I need to experiment with the modules turned off yet
- I'm not aware of reason for changes in `foreman-installer`, do you have any specific concerns?
- still not sure about `$previous -> $new_value` check my comment above

ares commented on lib/kafo/scenario_manager.rb:
```diff
@@ -29,17 +29,18 @@ def available_scenarios
+        use = (File.expand_path(config_file) == @previous_scenario ? 'INSTALLED' : "use: --scenario #{scenario}")
+        say ::HighLine.color("  #{content[:name]} ", :title)
+        say "(#{use})"
+        say "        " + content[:description] unless content[:description].nil? || content[:description].empty?
```
please don't combine `unless` and `||`, I'd prefer `if` with two negations

ares commented on lib/kafo/kafo_configure.rb:
```diff
@@ -39,6 +40,7 @@ def hooking
     end
 
     def initialize(*args)
+      self.class.preset_color_scheme
```
since this happens before the logger is built, could you make the color scheme used at https://github.com/theforeman/kafo/blob/master/lib/kafo/logger.rb#L95 so we don't duplicate ARGV parsing logic?

mbacovsky commented:
@ares: I fixed things from your comments. I also updated the way how params are diffed and took module "enabledness" into account. Check the comments in the last two commit logs for more details.

ares commented:
except one nit looks good to me, could you squash? I'd test and merge accordingly tomorrow

mbacovsky commented:
@ares, thanks for prompt review. Squashed, nit fixed.

ares commented on lib/kafo/scenario_manager.rb:
```diff
@@ -98,23 +116,62 @@ def scenario_from_args
+      say "\n" + ::HighLine.color("Overview of modules used in the scenarios (#{prev_conf.app[:name]} -> #{new_conf.app[:name]}):", :title)
+      modules = Hash.new { |h, k| h[k] = {} }
+      modules = prev_conf.modules.inject(modules) { |mods, mod| mods[mod.name][:prev] = mod.enabled?; mods }
+      modules = new_conf.modules.inject(modules) { |mods, mod| mods[mod.name][:new] = mod.enabled?; mods }
```
`:prev` and `:new` is much better :-)

ares commented:
Thanks @mbacovsky, merging!

mbacovsky commented:
Thanks @ares, very much appreciated :beers:!
